### DISCUSSION TITLE
get/setSelectedText - became very slow

### DISCUSSION THREAD

**Discussion by hakubo** [johnlindquist/kit#685] [Errors]
I need to wait about a second for each to complete.
I have reinstalled Kit and I'm on version 1.5.1.

https://user-images.githubusercontent.com/1018759/149298752-41b36391-1e55-4247-b648-b4425bb45e14.mov

**Comment by johnlindquist** [comment]
Dang, sorry you're dealing with this. I'm guessing it's something to do with how `osascript` AKA "applescript" is being run and not those specific functions. `getSelectedText/setSelectedText` really don't do much. Here's their entire source:

https://github.com/johnlindquist/kit/blob/main/src/lib/text.ts

## Try Hard-coding Path to `osascript`

### 1. Change osascript in kit
Open: `~/.kit/platform/darwin.js`

Find: `osascript`
Change it to `/usr/bin/osascript`

### 2. Clear cached/idle processes

This will sound strange, but open/close the Kit.app prompt 3+ times. Kit.app keeps 3 idle processes that have cached the previous version of `darwin.js`, so if you try to run your script now, you won't see the change

### 3. Run Your Script and Let Me Know If That Helps 😇

## Why?

Sometimes a tool on the PATH, often `nvm`, slows down the lookup of other tools. Hard-coding the full path skips the lookup. If this works for you, I'll hard-code it in the next release.

## If That's Still Slow, Please Compare These 2 Scripts 🙏

### Test 1 - Using `exec`

This will use `exec` rather than the built-in [applescript](https://github.com/johnlindquist/kit/blob/main/src/platform/darwin.ts#L6)

```js
// Name: Testing Selected Using Exec

import "@johnlindquist/kit"

// "getSelectedText"
hide()
let command = `tell application "System Events" to keystroke "c" using command down`
await exec(`osascript '-e' '${command}'`);

// Do something with text
let reversed = (await paste()).split("").reverse().join("")

// "setSelectedText"
command = `set the clipboard to "${reversed}"`
await exec(`osascript '-e' '${command}'`);

command = `tell application "System Events" to keystroke "v" using command down`
await exec(`osascript '-e' '${command}'`);
```

### Test 2 - Using `applescript` helper

This will use kit's [applescript](https://github.com/johnlindquist/kit/blob/main/src/platform/darwin.ts#L6) wrapper

```js
// Name:Testing Selecting Using AppleScript

import "@johnlindquist/kit"

"getSelectedText"
hide()
let command = `tell application "System Events" to keystroke "c" using command down`
await applescript(command);

// do something with text
let reversed = (await paste()).split("").reverse().join("")

// "setSelectedText"
command = `set the clipboard to "${reversed}"`
await applescript(command);

command = `tell application "System Events" to keystroke "v" using command down`
await applescript(command);
```

## Results?

If both of these scripts are slow, then something on your machine is causing `osascript` to slow down. I really hope it doesn't get this far, because I would be at a loss at this point 😬. We would have to try running `osascript` from the terminal directly and see if that helps.

**Comment by hakubo** [comment]
I'm still fighting with this and tried different things including https://github.com/johnlindquist/kit/pull/727

but when I compare osascript ran in terminal to the one via Script Kit I get like 2s slowdown :(

**Comment by hakubo** [comment]
anyway this is my workaround for now: 

```
let item = await arg(
  "Hit enter to paste",
  (await getClipboardHistory()).map((item) => ({
    ...item,
    value: { id: item.id, value: item.value },
  }))
);

removeClipboardItem(item.id);
await copy(item.value);
```

removeClipboardItem - is make sure I don't get duplicated entries on paste

this does mean I have to paste manually but at least I don't have to wait 2s

**Reply by johnlindquist** [reply]
Did you try the full path to osascript from the comments above? Not using the full path is almost always the cause of slowdowns with commands

**Reply by hakubo** [reply]
I did. No perf change.

My .zshrc imports few things and can take a second or two, could that be this?

**Reply by johnlindquist** [reply]
I think that should only happen if your script is sourcing that profile...

Hmm, you down for a Zoom next week? I'll dm you on twitter

**Comment by jacob-shuman** [comment]
I'm experiencing this same issue although it feels as if `getSelectedText` is even slower for me and `setSelectedText`doesn't seem to work at all 😭 

Please let me know if theres anything I can do to help debug this!

**Reply by jacob-shuman** [reply]
Not sure if this is relevant:

When i run the following line:
```ts
await setSelectedText("cool");
```

it creates a `latest.scpt` with the following content and clears my clipboard:
```
set the clipboard to ""
```

Is this expected? I didn't think setSelectedText would clear my clipboard?

**Reply by johnlindquist** [reply]
@jacob-shuman if it's not working at all, it's probably a permissions issue. 

The behavior of `setSelectedText` is:

1. Hide the UI so your script can paste to the app behind it
2. Write the string to the clipboard
3. Press a "cmd+v" to paste the contents from the clipboard:
4. Clear the clipboard as a cleanup step (this is why you're seeing that file)

```js
global.setSelectedText = async text => {
  await global.hide()

  await applescript(
    String.raw`set the clipboard to "${text.replaceAll(
      '"',
      '\\"'
    )}"`
  )

  await applescript(
    String.raw`tell application "System Events" to keystroke "v" using command down`
  )

  await applescript(String.raw`set the clipboard to ""`)
}
```

The "cmd+v" step requires security permissions:

<img width="721" alt="CleanShot 2022-03-21 at 08 15 25@2x" src="https://user-images.githubusercontent.com/36073/159280054-06fdd9a1-23f2-496a-bec2-c01add38f416.png">

**Reply by jacob-shuman** [reply]
That would explain it 🤦🏻‍♂️  it looks like I forgot to enable the accessibility permissions, the issue was fixed immediately after doing that.

Thank you so much for the detailed response! I've been really enjoying ScriptKit!
